language-plutus-core/plutus-ir/Language/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -0,0 +1,615 @@
+         -> TypeCheckM uni ann a
+         -> TypeCheckM uni ann a
+withBind b m = case b of
+  TermBind _ _ (VarDecl _ n ty) _rhs -> do
+      -- OPTIMIZE: redundant, usually this function is called together with `checkWellformBind`,
+      -- which performs normalization there as well
+      vTy <- normalizeTypeM $ void ty
+      withVar n vTy m
+  TypeBind _ (TyVarDecl _ tn k) _rhs ->
+      withTyVar tn (void k) m
+  DatatypeBind _ dt@(Datatype ann (TyVarDecl _ tn k) tyargs des vdecls) -> do
+      desType <- mkDestructorType
+      -- normalize all data-constructors's types.
+      normConstrs <- for vdecls $ \(VarDecl _ n ty) ->
+                                     -- make sure to add the implicitly-qualified tyargs foralls for each constructor.
+                                     (n,) <$> (normalizeTypeM $ void $ PIR.mkIterTyForall tyargs ty)
+      -- extend the type-environment with data-destructor+data-constructors
+      withVars ((des,desType):normConstrs) $
+          -- extend the kind-environment with type-constructor
+          withTyVar tn (void k) m
+   where
+       mkDestructorType :: TypeCheckM uni ann (Normalized (Type TyName uni ()))
+       mkDestructorType = do
+           -- get a fresh result type
+           out <- resultTypeName dt
+           normalizeTypeM $ void $
+             -- forall (a1::*) (a2::*) ... .
+             PIR.mkIterTyForall tyargs $
+               TyFun ann
+                     -- first argument is datatype:  T a1 a2 ... ->
+                     (PIR.mkIterTyApp ann (TyVar ann tn) (fmap (mkTyVar ann) tyargs ))
+                     -- forall (out :: *)
+                     (TyForall ann out (Type ann) $
+                       -- the constructors' types with result-type replaced by out
+                       -- note to self: no need to normalize the individual constructor here;
+                       -- the outer,final,surrounding normalizeTypeM will reach and normalize each constructor
+                       foldr
+                       (TyFun ann . constructorCaseType (TyVar ann out))
+                       -- result type of constructor is: out
+                       (TyVar ann out)
+                       vdecls
+                     )
+
+       withVars :: Foldable t
+                => t (Name, Normalized (Type TyName uni ()))
+                -> TypeCheckM uni ann a -> TypeCheckM uni ann a
+       withVars names m1 = foldr (uncurry withVar) m1 names
+
+-- | check that a binding is well-formed (correctly typed, kinded)
+checkWellformBind :: forall uni ann.
+                     (GShow uni, GEq uni, DefaultUni <: uni)
+                    => Recursivity
+                    -> Binding TyName Name uni ann
+                    -> TypeCheckM uni ann ()
+checkWellformBind recurs = \case
+  -- W-Term
+  TermBind _ _ (VarDecl ann _ ty) rhs -> do
+      checkKindM ann ty $ Type ()
+      -- OPTIMIZE: redundant, usually this function is called together with `withBind`,
```

Comment:
Tricky, I can't see a way to make that work without merging the functions.

Reply:
Yes, we could do that, but this way it mirrors the paper more closely, which is nice.

Reply:
The current setup seems to have a problem, see [this comment](https://github.com/input-output-hk/plutus/pull/2181#discussion_r459616549).